apptail/storage/tail_tracker.go

```diff
@@ -0,0 +1,231 @@
+package storage
+
+import (
+	"fmt"
+	"github.com/ActiveState/log"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type Tracker interface {
+	Update(instKey string, childKey string, childVal int64)
+	LoadTailers()
+	Remove(key string)
+	RegisterInstance(instKey string)
+	InitializeChildNode(instKey string, childkey string, offSet int64)
+	Commit() error
+	StartSubmissionTimer(persistInterval time.Duration)
+	IsInstanceRegistered(instKey string) bool
+	IsChildNodeInitialized(instKey string, childkey string) bool
+	GetFileCachedOffset(instkey string, fname string) int64
+	getBuffer() ([]byte, error)
+	CleanUp(validIds map[string]bool)
+}
+
+type BoxedInt64 struct{ V int64 }
+
+type TailNode map[string]*BoxedInt64
+
+type Tailer struct {
+	Instances map[string]TailNode
+}
+
+type tracker struct {
+	storage       Storage
+	Cached        *Tailer // do not expose this, it should ONLY be updated via Tracker methods
+	mux           *sync.Mutex
+	writeMux      *sync.Mutex
+	timerStopChan chan struct{} // used to send quit signal to time,
+	debug         bool
+}
+
+func NewTracker(s Storage, debug bool) Tracker {
+
+	return &tracker{
+		storage:  s,
+		mux:      &sync.Mutex{},
+		writeMux: &sync.Mutex{},
```

Comment:
It would be more
```go
type FileStorage struct {
    file_path string
    mux       *sync.Mutex
}
func (s *FileStorage) Write(buf []byte) error {
    s.mux.Lock()
    defer s.mux.Unlock()

    if err := ioutil.WriteFile(s.file_path, buf, FILE_MODE); err != nil {
        return err
    }

    // this extra step to make the file accessible by stackato user
    if err := os.Chmod(s.file_path, FILE_MODE); err != nil {
        return err
    }
    return nil
}
```
Where you're currently handling it in `Commit()`; this way, it's the `storage`'s responsibility to deal with, umm, storage, and `tracker` just needs to know to call it.